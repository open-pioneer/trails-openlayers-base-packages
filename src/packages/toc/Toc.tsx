// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { Box, Text, FormControl, FormLabel } from "@open-pioneer/chakra-integration";
import { FC } from "react";
import { CommonComponentProps, useCommonComponentProps } from "@open-pioneer/react-utils";
import { BasemapSwitcher, BasemapSwitcherProps } from "@open-pioneer/basemap-switcher";
import { useIntl } from "open-pioneer:react-hooks";

export interface TocProps extends CommonComponentProps {
    /**
     * The id of the map.
     */
    mapId: string;

    /**
     * Defines whether the basemap switcher is shown in the toc.
     * Defaults to false.
     */
    hideBasemapSwitcher?: boolean;

    /**
     * Properties for the embedded basemap switcher.
     * Property "mapId" is not applied.
     */
    basemapSwitcherProps?: Omit<BasemapSwitcherProps, "mapId">;
}

export const Toc: FC<TocProps> = (props) => {
    const intl = useIntl();

    const { mapId, hideBasemapSwitcher = false, basemapSwitcherProps } = props;
    const { containerProps } = useCommonComponentProps("toc", props);
    const basemapsLabel = intl.formatMessage({ id: "basemapsLabel" });
    const tocTitel = intl.formatMessage({ id: "tocTitel" });

    return (
        <Box {...containerProps}>
            {/*TODO: remove header as it should be added by an app developer if needed? */}
            <Box
                className="toc-header"
                padding={2}
                backgroundColor="var(--chakra-colors-blackAlpha-500)"
            >
                <Text as="b">{tocTitel}</Text>
            </Box>
            {hideBasemapSwitcher || (
                <Box className="toc-content" padding={2}>
                    <FormControl>
                        <FormLabel ps={1}>
                            <Text as="b">{basemapsLabel}:</Text>
                        </FormLabel>
                        <BasemapSwitcher {...basemapSwitcherProps} mapId={mapId}></BasemapSwitcher>
                    </FormControl>
                </Box>
            )}
        </Box>
    );
};
