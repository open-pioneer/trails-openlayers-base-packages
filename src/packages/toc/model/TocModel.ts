// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { createContext, Provider, useContext } from "react";

const TocModelContext = createContext<TocModel | undefined>(undefined);

/**
 * Provider for the toc's shared model.
 * Used at the root of the component.
 */
export const TocModelProvider: Provider<TocModel> = TocModelContext.Provider as Provider<TocModel>;

/**
 * Returns the shared model for the current toc component.
 */
export function useTocModel(): TocModel {
    const model = useContext(TocModelContext);
    if (!model) {
        throw new Error("Internal error: TocModel not found in context.");
    }
    return model;
}

/**
 * Shared model used by the toc and all its sub-components.
 *
 * @internal
 */
export interface TocModel {
    /**
     * Return the global widget options.
     *
     * NOTE: The object itself is reactive, but individual properties are not (change -> replace).
     */
    readonly options: TocWidgetOptions;

    /**
     * Returns the item that corresponds with the `layerId`.
     */
    getItem(layerId: string): TocItem | undefined;

    /**
     * Returns the list of all items.
     */
    getItems(): TocItem[];

    // Used by toc item components to register themselves
    registerItem(item: TocItem): void;
    unregisterItem(item: TocItem): void;
}

/**
 * Global toc widget options.
 *
 * @internal
 */
export interface TocWidgetOptions {
    /**
     * True: When showing a child, show all parents as well (`setVisible(true)`).
     */
    autoShowParents: boolean;

    /**
     * True: Layer items with children can be collapsed.
     */
    collapsibleGroups: boolean;

    /**
     * True: All groups are initially collapsed.
     */
    initiallyCollapsed: boolean;
}

/**
 * Represents an item in the toc.
 *
 * Currently items register themselves in the model when they are mounted
 * and remove themselves when they are unmounted.
 *
 * @internal
 */
export interface TocItem {
    /**
     * Identifier of the layer that corresponds with the list item.
     */
    readonly layerId: string;

    /**
     * true if list item is expanded.
     */
    readonly isExpanded: boolean;

    /**
     * Expands or collapses the list item.
     *
     * Note: not all list items support this operation.
     */
    setExpanded(expanded: boolean): void;
}
