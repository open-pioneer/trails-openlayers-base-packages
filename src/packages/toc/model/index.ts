// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
export { TocModelProvider, useTocModel } from "./context";
export { TocApiImpl } from "./TocApiImpl";
export { TocItemImpl } from "./TocItemImpl";
export { createOptions, TocModel, type TocWidgetOptions } from "./TocModel";
export {
    type ExpandItemOptions,
    type TocApi,
    type TocDisposedEvent,
    type TocDisposedHandler,
    type TocItem,
    type TocReadyEvent,
    type TocReadyHandler
} from "./types";
