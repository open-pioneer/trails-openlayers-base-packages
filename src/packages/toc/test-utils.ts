// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { screen } from "@testing-library/react";

export async function findToc() {
    const tocDiv = await screen.findByTestId("toc");
    return tocDiv;
}
