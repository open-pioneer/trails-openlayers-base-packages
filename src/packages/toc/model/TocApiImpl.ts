// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { TocModel } from "./TocModel";
import { TocApi, TocItem } from "./types";

export class TocApiImpl implements TocApi {
    #tocModel: TocModel;

    constructor(model: TocModel) {
        this.#tocModel = model;
    }

    getItemById(id: string): TocItem | undefined {
        return this.#tocModel.getItemById(id);
    }

    getItemByLayerId(layerId: string): TocItem | undefined {
        return this.#tocModel.getItemByLayerId(layerId);
    }

    getItems(): TocItem[] {
        return this.#tocModel.getItems();
    }
}
