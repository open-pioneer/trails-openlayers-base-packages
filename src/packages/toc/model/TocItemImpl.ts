// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { Reactive, reactive } from "@conterra/reactivity-core";
import { AnyLayer } from "@open-pioneer/map";
import { TocModel } from "./TocModel";
import { ExpandItemOptions, TocItem } from "./types";

export class TocItemImpl implements TocItem {
    #htmlElement = reactive<HTMLElement | undefined>();
    #expanded: Reactive<boolean>;
    #layer: AnyLayer;
    #tocModel: TocModel;

    constructor(layer: AnyLayer, tocModel: TocModel, expanded: boolean) {
        this.#expanded = reactive(expanded);
        this.#layer = layer;
        this.#tocModel = tocModel;
    }

    get id() {
        return this.layerId;
    }

    get layerId() {
        return this.#layer.id;
    }

    get isExpanded() {
        return this.#expanded.value;
    }

    get htmlElement() {
        return this.#htmlElement.value;
    }

    setExpanded(expanded: boolean, options?: ExpandItemOptions): void {
        this.#expanded.value = expanded;
        let bubble = options?.bubble;
        //by default bubble if expand is true
        if (bubble === undefined) {
            bubble = expanded;
        }

        if (bubble) {
            const parentLayer = this.#layer.parent;
            if (parentLayer) {
                this.#tocModel.getItemByLayerId(parentLayer.id)?.setExpanded(expanded, options);
            }
        }
    }

    //private setter, not exposed in TocItem interface
    setHtmlElement(element: HTMLElement | undefined) {
        this.#htmlElement.value = element;
    }
}
