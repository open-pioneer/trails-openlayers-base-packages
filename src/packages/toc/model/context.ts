// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { createContext, Provider, useContext } from "react";
import { TocModel } from "./TocModel";

const TocModelContext = createContext<TocModel | undefined>(undefined);

/**
 * Provider for the toc's shared model.
 * Used at the root of the component.
 */
export const TocModelProvider: Provider<TocModel> = TocModelContext.Provider as Provider<TocModel>;

/**
 * Returns the shared model for the current toc component.
 */
export function useTocModel(): TocModel {
    const model = useContext(TocModelContext);
    if (!model) {
        throw new Error("Internal error: TocModel not found in context.");
    }
    return model;
}
