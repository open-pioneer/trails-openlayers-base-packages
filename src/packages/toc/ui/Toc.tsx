// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { reactive, reactiveMap } from "@conterra/reactivity-core";
import { BasemapSwitcher, BasemapSwitcherProps } from "@open-pioneer/basemap-switcher";
import { Box, Flex, Spacer, Text } from "@open-pioneer/chakra-integration";
import { MapModel, MapModelProps, useMapModel } from "@open-pioneer/map";
import {
    CommonComponentProps,
    SectionHeading,
    TitledSection,
    useCommonComponentProps
} from "@open-pioneer/react-utils";
import { useIntl } from "open-pioneer:react-hooks";
import { FC, useEffect, useId, useMemo, useRef } from "react";
import { TocItem, TocModel, TocModelProvider, TocWidgetOptions } from "../model/TocModel";
import { TopLevelLayerList } from "./LayerList/LayerList";
import { Tools } from "./Tools";

/**
 * Props supported by the {@link Toc} component.
 */
export interface TocProps extends CommonComponentProps, MapModelProps {
    /**
     * Defines whether the tool component is shown in the toc.
     *
     * Defaults to `false`.
     */
    showTools?: boolean;

    /**
     * Properties for the embedded tool component.
     */
    toolsConfig?: ToolsConfig;

    /**
     * Defines whether the basemap switcher is shown in the toc.
     *
     * Defaults to `true`.
     */
    showBasemapSwitcher?: boolean;

    /**
     * Properties for the embedded basemap switcher.
     *
     * Property `mapId` is not applied (the basemap switcher uses the same map as the).
     */
    basemapSwitcherProps?: Omit<BasemapSwitcherProps, "mapId">;

    /**
     * If `true`, groups in the toc can be collapsed and expanded.
     * This property should only be `true` if the map actually contains layer groups.
     *
     * Defaults to `false`.
     */
    collapsibleGroups?: boolean;

    /**
     * If `true` groups in the toc are collapsed initially.
     *
     * Defaults to `false`. If {@link collapsibleGroups} is `false` this property should also be `false`. Otherwise only the top level layers will appear in the toc.
     */
    initiallyCollapsed?: boolean;

    /**
     * Show the parent layers when a child layer is made visible.
     *
     * Defaults to `true`.
     */
    autoShowParents?: boolean;
}

/**
 * Props supported by the {@link Tools} component.
 */
export interface ToolsConfig {
    /**
     * Optional property to show the `hide all layers` entry.
     *
     * Defaults to `true`.
     */
    showHideAllLayers?: boolean;

    /**
     * Optional property to show the `collapse all groups` entry.
     *
     * Defaults to `true`. Only applicable if {@link TocProps.collapsibleGroups} is `true`.
     */
    showCollapseAllGroups?: boolean;
}

const PADDING = 2;

/**
 * Displays the layers of the configured map.
 */
export const Toc: FC<TocProps> = (props: TocProps) => {
    const intl = useIntl();
    const { containerProps } = useCommonComponentProps("toc", props);
    const state = useMapModel(props);

    let content: JSX.Element | null;
    switch (state.kind) {
        case "loading":
            content = null;
            break;
        case "rejected":
            content = <Text className="toc-error">{intl.formatMessage({ id: "error" })}</Text>;
            break;
        case "resolved": {
            const map = state.map;
            content = <TocContent {...props} map={map} />;
            break;
        }
    }

    return (
        <Flex {...containerProps} direction="column" gap={PADDING}>
            {content}
        </Flex>
    );
};

/** This component is rendered once we have a reference to the loaded map model. */
function TocContent(props: TocProps & { map: MapModel }) {
    const {
        map,
        showTools = false,
        toolsConfig,
        showBasemapSwitcher = true,
        basemapSwitcherProps
    } = props;
    const intl = useIntl();
    const model = useTocModel(props);
    const basemapsHeadingId = useId();
    const basemapSwitcher = showBasemapSwitcher && (
        <Box className="toc-basemap-switcher">
            <TitledSection
                title={
                    <SectionHeading id={basemapsHeadingId} size={"sm"} mb={PADDING}>
                        {intl.formatMessage({ id: "basemapsLabel" })}
                    </SectionHeading>
                }
            >
                <BasemapSwitcher
                    map={map}
                    aria-labelledby={basemapsHeadingId}
                    {...basemapSwitcherProps}
                />
            </TitledSection>
        </Box>
    );

    const layerList = (
        <Box className="toc-operational-layers">
            <TitledSection
                title={
                    <SectionHeading size={"sm"} mb={2}>
                        <Flex>
                            <Text my={3}>
                                {intl.formatMessage({
                                    id: "operationalLayerLabel"
                                })}
                            </Text>
                            <Spacer />
                            {showTools && <Tools map={map} {...toolsConfig} />}
                        </Flex>
                    </SectionHeading>
                }
            >
                <TopLevelLayerList
                    map={map}
                    aria-label={intl.formatMessage({ id: "operationalLayerLabel" })}
                />
            </TitledSection>
        </Box>
    );

    return (
        <TocModelProvider value={model}>
            {basemapSwitcher}
            {layerList}
        </TocModelProvider>
    );
}

function useTocModel(props: TocProps): TocModel {
    const initialProps = useRef(props);
    const { model, options } = useMemo(() => {
        const options = reactive<TocWidgetOptions>(
            createOptions(
                initialProps.current.autoShowParents,
                initialProps.current.collapsibleGroups,
                initialProps.current.initiallyCollapsed
            )
        );

        // Indexed by layerId
        const items = reactiveMap<string, TocItem>();
        const model: TocModel = {
            get options() {
                return options.value;
            },
            getItem(layerId: string): TocItem | undefined {
                return items.get(layerId);
            },
            getItems(): TocItem[] {
                return Array.from(items.values());
            },
            registerItem(item: TocItem): void {
                if (items.has(item.layerId)) {
                    throw new Error(`Item with layerId '${item.layerId}' already registered.`);
                }
                items.set(item.layerId, item);
            },
            unregisterItem(item: TocItem): void {
                if (items.get(item.layerId) !== item) {
                    throw new Error(`Item with layerId '${item.layerId}' not registered.`);
                }
                items.delete(item.layerId);
            }
        };
        return { model, options };
    }, []);

    // Sync props to model
    useEffect(() => {
        options.value = createOptions(
            props.autoShowParents,
            props.collapsibleGroups,
            props.initiallyCollapsed
        );
    }, [options, props.autoShowParents, props.collapsibleGroups, props.initiallyCollapsed]);

    return model;
}

function createOptions(
    autoShowParents?: boolean | undefined,
    collapsibleGroups?: boolean | undefined,
    isCollapsed?: boolean | undefined
): TocWidgetOptions {
    return {
        autoShowParents: autoShowParents ?? true,
        collapsibleGroups: collapsibleGroups ?? false,
        initiallyCollapsed: isCollapsed ?? false
    };
}
