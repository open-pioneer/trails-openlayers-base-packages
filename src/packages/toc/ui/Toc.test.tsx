// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { createServiceOptions, setupMap } from "@open-pioneer/map-test-utils";
import { PackageContextProvider } from "@open-pioneer/test-utils/react";
import { act, render, screen, waitFor, fireEvent } from "@testing-library/react";
import TileLayer from "ol/layer/Tile";
import { expect, it, vi } from "vitest";
import { Toc } from "./Toc";
import { TocDisposedEvent, TocReadyEvent } from "../model/TocModel";
import { SimpleLayer, GroupLayer } from "@open-pioneer/map";

const BASEMAP_SWITCHER_CLASS = ".basemap-switcher";
const BASEMAP_SWITCHER_SELECT_CLASS = ".basemap-switcher-select";

it("should successfully create a toc component", async () => {
    const { map, registry } = await setupMap({
        layers: [
            {
                title: "Base layer",
                id: "base-layer",
                olLayer: new TileLayer({}),
                isBaseLayer: true
            },
            {
                title: "Layer 1",
                id: "layer-1",
                olLayer: new TileLayer({})
            },
            {
                title: "Layer 2",
                id: "layer-2",
                olLayer: new TileLayer({})
            }
        ]
    });
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Toc map={map} data-testid="toc" />
        </PackageContextProvider>
    );

    const tocDiv = await findToc();
    const { basemapSelect } = await waitForBasemapSwitcher(tocDiv!);

    // react-select creates list of options in dom after opening selection
    act(() => {
        fireEvent.keyDown(basemapSelect, { key: "ArrowDown" });
    });

    await waitFor(() => {
        const options = tocDiv.getElementsByClassName("basemap-switcher-option");
        if (options.length !== 1 || options[0]?.textContent !== "Base layer") {
            throw new Error("expected basemap switcher to contain the Base layer option");
        }
    });
    expect(tocDiv).toMatchSnapshot();
});

it("should successfully create a toc component with additional css classes", async () => {
    const { map, registry } = await setupMap();
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Toc map={map} className="test" data-testid="toc" />
        </PackageContextProvider>
    );

    const tocDiv = await findToc();
    expect(tocDiv.classList.contains("test")).toBe(true);
});

it("should embed the basemap switcher by default", async () => {
    const { map, registry } = await setupMap();
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Toc map={map} data-testid="toc"></Toc>
        </PackageContextProvider>
    );

    const tocDiv = await findToc();
    const { basemapSwitcher } = await waitForBasemapSwitcher(tocDiv!);
    expect(basemapSwitcher.tagName).toBe("DIV");
});

it("should not show the basemap switcher if 'showBasemapSwitcher' is set to false", async () => {
    const { map, registry } = await setupMap();
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Toc map={map} showBasemapSwitcher={false} data-testid="toc"></Toc>
        </PackageContextProvider>
    );

    const tocDiv = await findToc();
    const basemapSwitcher = tocDiv.querySelector(BASEMAP_SWITCHER_CLASS);
    expect(basemapSwitcher).toBeNull();
});

it("should support overriding basemap-switcher properties", async () => {
    const { map, registry } = await setupMap({
        layers: [
            {
                title: "OSM",
                olLayer: new TileLayer({}),
                isBaseLayer: true
            }
        ]
    });

    const injectedServices = createServiceOptions({ registry });
    render(
        <PackageContextProvider services={injectedServices}>
            <Toc
                map={map}
                basemapSwitcherProps={{
                    allowSelectingEmptyBasemap: true,
                    className: "test-class"
                }}
                data-testid="toc"
            />
        </PackageContextProvider>
    );

    const tocDiv = await findToc();
    const { basemapSwitcher, basemapSelect } = await waitForBasemapSwitcher(tocDiv!);
    expect(basemapSwitcher?.classList.contains("test-class")).toBe(true);

    // react-select creates list of options in dom after opening selection
    act(() => {
        fireEvent.keyDown(basemapSelect, { key: "ArrowDown" });
    });

    await waitFor(() => {
        const options = basemapSelect.getElementsByClassName("basemap-switcher-option");
        const optionLabels = Array.from(options).map((opt) => opt.textContent);
        expect(optionLabels, "basemap options are not equal to their expected values")
            .toMatchInlineSnapshot(`
        [
          "OSM",
          "emptyBasemapLabel",
        ]
      `);
    });
});

it("should raise onReady and onDispose events", async () => {
    const { map, registry } = await setupMap({
        layers: [
            {
                title: "Base layer",
                id: "base-layer",
                olLayer: new TileLayer({}),
                isBaseLayer: true
            },
            {
                title: "Layer 1",
                id: "group",
                olLayer: new TileLayer({})
            },
            {
                title: "Layer 2",
                id: "layer-2",
                olLayer: new TileLayer({})
            }
        ]
    });
    const injectedServices = createServiceOptions({ registry });

    let readyEvent: TocReadyEvent | undefined;
    const onReadyHandler = (e: TocReadyEvent) => {
        readyEvent = e;
    };
    let disposedEvent: TocDisposedEvent | undefined;
    const onDisposedHandler = (e: TocDisposedEvent) => {
        disposedEvent = e;
    };

    const onReadyMock = vi.fn().mockImplementation(onReadyHandler);
    const onDisposedMock = vi.fn().mockImplementation(onDisposedHandler);

    const { unmount } = render(
        <PackageContextProvider services={injectedServices}>
            <Toc map={map} data-testid="toc" onDisposed={onDisposedMock} onReady={onReadyMock} />
        </PackageContextProvider>
    );
    await findToc();

    await waitFor(() => {
        expect(onReadyMock).toBeCalled();
    });
    expect(readyEvent).toBeDefined();
    expect(readyEvent?.api).toBeDefined();

    unmount(); //unmount toc

    await waitFor(() => {
        expect(onDisposedMock).toBeCalled();
    });
    expect(disposedEvent).toBeDefined();

    //should have been called exactly once
    expect(onReadyMock).toBeCalledTimes(1);
    expect(onDisposedMock).toBeCalledTimes(1);
});

it("should return TocItems for layers via Toc API", async () => {
    const { map, registry } = await setupMap({
        layers: [
            {
                title: "Base layer",
                id: "base-layer",
                olLayer: new TileLayer({}),
                isBaseLayer: true
            },
            {
                title: "Layer 1",
                id: "layer-1",
                olLayer: new TileLayer({})
            },
            {
                title: "Layer 2",
                id: "layer-2",
                olLayer: new TileLayer({})
            }
        ]
    });
    const injectedServices = createServiceOptions({ registry });

    let readyEvent: TocReadyEvent | undefined;
    const onReadyHandler = (e: TocReadyEvent) => {
        readyEvent = e;
    };
    const onReadyMock = vi.fn().mockImplementation(onReadyHandler);

    render(
        <PackageContextProvider services={injectedServices}>
            <Toc map={map} data-testid="toc" onReady={onReadyMock} />
        </PackageContextProvider>
    );
    await findToc();

    await waitFor(() => {
        expect(onReadyMock).toBeCalled();
    });
    expect(readyEvent).toBeDefined();
    expect(readyEvent?.api).toBeDefined();

    const api = readyEvent?.api;
    const tocItems = api?.getItems();
    expect(tocItems?.length).toBe(2); //only two items because one layer is a base alyer
    expect(api?.getItemByLayerId("base-layer")).toBeUndefined();
    expect(api?.getItemByLayerId("layer-1")).toBeDefined();
    expect(api?.getItemByLayerId("layer-2")).toBeDefined();
});

it("should provide access to LayerItem HTMLElement via Toc API", async () => {
    const { map, registry } = await setupMap({
        layers: [
            {
                title: "Base layer",
                id: "base-layer",
                olLayer: new TileLayer({}),
                isBaseLayer: true
            },
            {
                title: "Layer 1",
                id: "layer-1",
                olLayer: new TileLayer({})
            },
            {
                title: "Layer 2",
                id: "layer-2",
                olLayer: new TileLayer({})
            }
        ]
    });
    const injectedServices = createServiceOptions({ registry });

    let readyEvent: TocReadyEvent | undefined;
    const onReadyHandler = (e: TocReadyEvent) => {
        readyEvent = e;
    };

    const onReadyMock = vi.fn().mockImplementation(onReadyHandler);
    const onDisposedMock = vi.fn();

    const { unmount } = render(
        <PackageContextProvider services={injectedServices}>
            <Toc map={map} data-testid="toc" onDisposed={onDisposedMock} onReady={onReadyMock} />
        </PackageContextProvider>
    );
    await findToc();

    await waitFor(() => {
        expect(onReadyMock).toBeCalled();
    });
    expect(readyEvent).toBeDefined();
    expect(readyEvent?.api).toBeDefined();

    const api = readyEvent?.api;
    const tocItem = api?.getItemByLayerId("layer-1");
    expect(tocItem).toBeDefined();

    const htmlElement = tocItem?.htmlElement;
    expect(htmlElement).toBeTruthy();
    expect(htmlElement?.classList).toContain("toc-layer-item");

    unmount(); //unmount toc;

    await waitFor(() => {
        expect(onDisposedMock).toBeCalled();
    });

    expect(tocItem?.htmlElement).toBeUndefined();
});

it("should toggle LayerItem via Toc API", async () => {
    const group = new GroupLayer({
        id: "group",
        title: "a group layer",
        visible: false,
        layers: [
            new GroupLayer({
                id: "subgroup",
                title: "a nested group layer",
                visible: false,
                layers: [
                    new SimpleLayer({
                        id: "submember",
                        title: "subgroup member",
                        olLayer: new TileLayer({}),
                        visible: false
                    })
                ]
            })
        ]
    });
    const { map, registry } = await setupMap({
        layers: [group]
    });
    const injectedServices = createServiceOptions({ registry });
    let readyEvent: TocReadyEvent | undefined;
    const onReadyHandler = (e: TocReadyEvent) => {
        readyEvent = e;
    };

    const onReadyMock = vi.fn().mockImplementation(onReadyHandler);

    render(
        <PackageContextProvider services={injectedServices}>
            <Toc map={map} data-testid="toc" onReady={onReadyMock} initiallyCollapsed={true} />
        </PackageContextProvider>
    );
    await findToc();

    await waitFor(() => {
        expect(onReadyMock).toBeCalled();
    });
    expect(readyEvent).toBeDefined();
    expect(readyEvent?.api).toBeDefined();

    const api = readyEvent?.api;
    const tocItemSubgroup = api?.getItemByLayerId("subgroup");
    expect(tocItemSubgroup).toBeDefined();
    expect(tocItemSubgroup?.isExpanded).toBeFalsy();
    const tocItemGroup = api?.getItemByLayerId("group");
    expect(tocItemGroup).toBeDefined();
    expect(tocItemGroup?.isExpanded).toBeFalsy();

    tocItemSubgroup?.setExpanded(true); //initially collapsed
    expect(tocItemSubgroup?.isExpanded).toBeTruthy();
    expect(tocItemGroup?.isExpanded).toBeTruthy(); //should expand parent by default

    tocItemSubgroup?.setExpanded(false);
    expect(tocItemSubgroup?.isExpanded).toBeFalsy();
    expect(tocItemGroup?.isExpanded).toBeTruthy(); //should not collapse parent by default

    tocItemSubgroup?.setExpanded(false, { bubble: true });
    expect(tocItemSubgroup?.isExpanded).toBeFalsy();
    expect(tocItemGroup?.isExpanded).toBeFalsy(); //should collapse parent as well if bubble is explicetly true

    tocItemSubgroup?.setExpanded(true, { bubble: false });
    expect(tocItemSubgroup?.isExpanded).toBeTruthy();
    expect(tocItemGroup?.isExpanded).toBeFalsy(); //should not expand parent as well if bubble is false
});

async function findToc() {
    return await screen.findByTestId("toc");
}

async function waitForBasemapSwitcher(tocDiv: HTMLElement) {
    return await waitFor(() => {
        const basemapSwitcher = tocDiv.querySelector(BASEMAP_SWITCHER_CLASS);
        if (!basemapSwitcher) {
            throw new Error("basemap switcher not mounted");
        }

        const basemapSelect = basemapSwitcher?.querySelector(BASEMAP_SWITCHER_SELECT_CLASS);
        if (!basemapSelect) {
            throw new Error("failed to find select element in basemap switcher");
        }
        return { basemapSwitcher, basemapSelect };
    });
}
