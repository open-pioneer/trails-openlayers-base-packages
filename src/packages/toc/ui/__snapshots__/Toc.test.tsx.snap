// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should successfully create a toc component 1`] = `
<div
  class="toc css-1op6ogi"
  data-testid="toc"
>
  <div
    class="toc-basemap-switcher css-c1sqso"
  >
    <h1
      class="chakra-heading css-k5qncm"
      id="«r0»"
    >
      basemapsLabel
    </h1>
    <div
      class="basemap-switcher css-0"
    >
      <div
        class="basemap-switcher-select chakra-select__root css-3jlqt6"
        data-part="root"
        data-scope="select"
        dir="ltr"
        id="select:«r1»"
      >
        <div
          class="chakra-select__control css-17j1053"
          data-part="control"
          data-scope="select"
          data-state="open"
          dir="ltr"
          id="select:«r1»:control"
        >
          <button
            aria-controls="select:«r1»:content"
            aria-expanded="true"
            aria-haspopup="listbox"
            aria-labelledby="«r0»"
            class="basemap-switcher-select-trigger chakra-select__trigger css-jjrwqe"
            data-part="trigger"
            data-placement="bottom-start"
            data-scope="select"
            data-state="open"
            dir="ltr"
            id="select:«r1»:trigger"
            role="combobox"
            type="button"
          >
            <span
              class="chakra-select__valueText css-176t9sc"
              data-part="value-text"
              data-scope="select"
              dir="ltr"
            >
              Base layer
            </span>
          </button>
          <div
            class="chakra-select__indicatorGroup css-1fx1ifu"
          >
            <div
              aria-hidden="true"
              class="chakra-select__indicator css-1atg1ut"
              data-part="indicator"
              data-scope="select"
              data-state="open"
              dir="ltr"
            >
              <svg
                class="css-s3mb0o"
                viewBox="0 0 24 24"
              >
                <path
                  d="m6 9 6 6 6-6"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="toc-operational-layers css-0"
  >
    <h1
      class="chakra-heading css-12sq6rj"
    >
      <div
        class="css-k008qs"
      >
        <p
          class="css-0"
        >
          operationalLayerLabel
        </p>
        <div
          class="css-xsawf1"
        />
      </div>
    </h1>
    <ul
      aria-label="operationalLayerLabel"
      class="toc-layer-list chakra-list__root css-1qja9vd"
      role="list"
    >
      <li
        class="toc-layer-item layer-layer-2 css-0"
      >
        <div
          class="toc-layer-item-content css-1rikyh3"
        >
          <label
            aria-label="Layer 2"
            class="chakra-checkbox__root css-uwx0lo"
            data-part="root"
            data-scope="checkbox"
            data-state="checked"
            dir="ltr"
            for="checkbox:«r3»:input"
            id="checkbox:«r3»"
          >
            <input
              aria-labelledby="checkbox:«r3»:label"
              checked=""
              id="checkbox:«r3»:input"
              style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap; word-wrap: normal;"
              type="checkbox"
              value="on"
            />
            <div
              aria-hidden="true"
              class="chakra-checkbox__control css-gjxnkp"
              data-part="control"
              data-scope="checkbox"
              data-state="checked"
              dir="ltr"
              id="checkbox:«r3»:control"
            >
              <svg
                class="css-wuqtn7"
                data-state="checked"
                viewBox="0 0 24 24"
              >
                <polyline
                  points="20 6 9 17 4 12"
                />
              </svg>
            </div>
            <span
              class="chakra-checkbox__label css-1td7ckk"
              data-part="label"
              data-scope="checkbox"
              data-state="checked"
              dir="ltr"
              id="checkbox:«r3»:label"
            >
              Layer 2
            </span>
          </label>
          <div
            class="css-xsawf1"
          />
        </div>
      </li>
      <li
        class="toc-layer-item layer-layer-1 css-0"
      >
        <div
          class="toc-layer-item-content css-1rikyh3"
        >
          <label
            aria-label="Layer 1"
            class="chakra-checkbox__root css-uwx0lo"
            data-part="root"
            data-scope="checkbox"
            data-state="checked"
            dir="ltr"
            for="checkbox:«r6»:input"
            id="checkbox:«r6»"
          >
            <input
              aria-labelledby="checkbox:«r6»:label"
              checked=""
              id="checkbox:«r6»:input"
              style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap; word-wrap: normal;"
              type="checkbox"
              value="on"
            />
            <div
              aria-hidden="true"
              class="chakra-checkbox__control css-gjxnkp"
              data-part="control"
              data-scope="checkbox"
              data-state="checked"
              dir="ltr"
              id="checkbox:«r6»:control"
            >
              <svg
                class="css-wuqtn7"
                data-state="checked"
                viewBox="0 0 24 24"
              >
                <polyline
                  points="20 6 9 17 4 12"
                />
              </svg>
            </div>
            <span
              class="chakra-checkbox__label css-1td7ckk"
              data-part="label"
              data-scope="checkbox"
              data-state="checked"
              dir="ltr"
              id="checkbox:«r6»:label"
            >
              Layer 1
            </span>
          </label>
          <div
            class="css-xsawf1"
          />
        </div>
      </li>
    </ul>
  </div>
</div>
`;
